torch/testing/asserts.py

```diff
@@ -0,0 +1,188 @@
+import functools
+import inspect
+import warnings
+from typing import Callable, Optional, Tuple
+
+import torch
+from torch.testing._internal.common_utils import get_comparison_dtype as _get_comparison_dtype, TestCase as _TestCase
+
+__all__ = ["assert_tensors_equal", "assert_tensors_allclose"]
+
+try:
+    import pytest
+
+    UsageError = pytest.UsageError
+except ImportError:
+
+    class UsageError(Exception):
+        pass
+
+
+class UsageWarning(Warning):
+    pass
+
+
+def _get_default_rtol_and_atol(a: torch.Tensor, b: torch.Tensor) -> Tuple[float, float]:
+    return _TestCase()._getDefaultRtolAndAtol(a.dtype, b.dtype)
+
+
+def _hide_internal_traceback(fn):
+    @functools.wraps(fn)
+    def wrapper(*args, **kwargs):
+        try:
+            return fn(*args, **kwargs)
+        except AssertionError:
+            for frame_info in inspect.trace():
+                frame_info.frame.f_locals.setdefault("__tracebackhide__", True)
+            raise
+
+    return wrapper
+
+
+# These attributes will be checked by default in every call to assert_tensors_(equal|allclose) and
+# assert_(equal|allclose) if the inputs are tensors.
+DEFAULT_CHECKED_ATTRIBUTES = (
+    "shape",
+    "dtype",
+)
+
+
+def _compare_tensors_meta(a: torch.Tensor, b: torch.Tensor, **attributes: bool) -> None:
+    if not (isinstance(a, torch.Tensor) and isinstance(b, torch.Tensor)):
+        raise AssertionError(f"Both inputs have to be tensors, but got {type(a)} and {type(b)} instead.")
+
+    for attr in DEFAULT_CHECKED_ATTRIBUTES:
+        attributes.setdefault(attr, True)
+
+    for attr, check in attributes.items():
+        if not check:
+            continue
+
+        try:
+            val_a = getattr(a, attr)
+            val_b = getattr(b, attr)
+            if val_a != val_b:
+                raise AssertionError(f"The values for attribute '{attr}' do not match: {val_a} != {val_b}.")
+        except AttributeError as error:
+            raise UsageError(f"At least one of the inputs does not have an attribute '{attr}'.") from error
+
+
+def _compare_tensors(
+    a: torch.Tensor,
+    b: torch.Tensor,
+    asserter: Callable[[torch.Tensor, torch.Tensor], None],
+    **strict_attributes: bool,
+) -> None:
+    if any(tensor.dtype in (torch.complex32, torch.complex64, torch.complex128) for tensor in (a, b)):
+        raise UsageError("Comparison for complex tensors is not supported yet.")
+    if any(tensor.is_quantized for tensor in (a, b)):
+        raise UsageError("Comparison for quantized tensors is not supported yet.")
+    if any(tensor.is_sparse for tensor in (a, b)):
+        raise UsageError("Comparison for sparse tensors is not supported yet.")
+
+    _compare_tensors_meta(a, b, **strict_attributes)
+
+    if a.device != b.device:
+        a = a.cpu()
+        b = b.cpu()
```

Comment:
Isn't this an implementation detail of `torch.eq` or `torch.all_close` ?

Reply:
No. PyTorch operations don't support implicit device movement. Our current assertEqual() in the test suite will automatically put tensors on the CPU, however, which can be convenient. It has also led to errors when a tensor on the wrong device was returned and silently coerced to a different device for equality checking.

My suggestion is to require the devices match by default.